daffodil-runtime1-unparser/src/main/scala/edu/illinois/ncsa/daffodil/processors/unparsers/PackedDecimalUnparsers.scala

```diff
@@ -0,0 +1,138 @@
+import java.lang.{ Number => JNumber, Long => JLong }
+import java.math.{ BigDecimal => JBigDecimal }
+import edu.illinois.ncsa.daffodil.exceptions.Assert
+import edu.illinois.ncsa.daffodil.processors.parsers.HasRuntimeExplicitLength
+import edu.illinois.ncsa.daffodil.processors.Evaluatable
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.LengthUnits
+import edu.illinois.ncsa.daffodil.schema.annotation.props.gen.YesNo
+import edu.illinois.ncsa.daffodil.io.DataOutputStream
+
+abstract class PackedNumberBaseUnparser(e: ElementRuntimeData)
+  extends PrimUnparserObject(e) {
+
+  protected def getBitLength(s: ParseOrUnparseState): Int
+
+  protected def putNumber(dos: DataOutputStream, number: JNumber, nBits: Int): Boolean
+
+  def unparse(state: UState): Unit = {
+    val nBits = getBitLength(state)
+    val node = state.currentInfosetNode.asSimple
+    val value = node.dataValue.asInstanceOf[JNumber]
+    val dos = state.dataOutputStream
+
+    val res =
+      if (nBits > 0) {
+        putNumber(dos, value, nBits)
+      } else {
+        true
+      }
+
+    if (!res) {
+      Assert.invariant(dos.maybeRelBitLimit0b.isDefined)
+      UnparseError(One(state.schemaFileLocation), One(state.currentLocation), "Insufficient space to unparse element %s, required %s bits, but only %s were available.",
+        e.dpathElementCompileInfo.namedQName.toPrettyString, nBits, dos.maybeRelBitLimit0b.get)
+    }
+  }
+
+}
+
+abstract class PackedIntegerBaseUnparser(e: ElementRuntimeData, signed: Boolean)
+  extends PackedNumberBaseUnparser(e) {
+
+  override def putNumber(dos: DataOutputStream, value: JNumber, nBits: Int): Boolean = {
+    dos.putByteArray(packedFromBigDecimal(new JBigDecimal(value.toString), nBits), nBits)
+  }
+}
+
+class PackedIntegerKnownLengthUnparser(e: ElementRuntimeData, signed: Boolean, override val lengthInBits: Int)
+  extends PackedIntegerBaseUnparser(e, signed)
+  with HasKnownLengthInBits {
+}
+
+class PackedIntegerRuntimeLengthUnparser(val e: ElementRuntimeData, signed: Boolean, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends PackedIntegerBaseUnparser(e, signed)
+  with HasRuntimeExplicitLength {
+
+  override val runtimeDependencies = List(lengthEv)
+}
+
+class PackedDecimalKnownLengthUnparser(e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int, val lengthInBits: Int)
+  extends PackedDecimalBaseUnparser(e, signed, binaryDecimalVirtualPoint)
+  with HasKnownLengthInBits {
+}
+
+class PackedDecimalRuntimeLengthUnparser(val e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends PackedDecimalBaseUnparser(e, signed, binaryDecimalVirtualPoint)
+  with HasRuntimeExplicitLength {
+
+  override val runtimeDependencies = List(lengthEv)
+}
+
+abstract class PackedDecimalBaseUnparser(e: ElementRuntimeData, signed: YesNo, binaryDecimalVirtualPoint: Int)
+  extends PackedNumberBaseUnparser(e) {
```

Comment:
signed and binaryDecimalVirtualPoint do not seem to be used (same with other new parsers). Are they needed, or are these properties not handled correctly? Maybe related, it doesn't look like the dfdl:binaryPackedSignCodes is implemented here. That should probably be implemented as part of this changeset as well.

Reply:
If these are needed for test cases like the formats on DFDLSchemas, or some of the IBM-provided tests, then we need to implement them. If not, we could leave these as NYI for now and implement later.